File: semestru 4/MotoProjectCs/server/Service.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using moto.model;
using moto.persistence;
using moto.services;
using log4net;
using moto.networking.jsonprotocol;

namespace moto.server
{
    public class Service : IService
    {
        private static readonly ILog log = LogManager.GetLogger(typeof(Service));
        private readonly ConcurrentDictionary<int, IObserver> loggedClients;
        private readonly IPlayerRepository playerRepository;
        private readonly IUserRepository userRepository;
        private readonly IRaceRepository raceRepository;
        private readonly ITeamRepository teamRepository;
        private readonly int defaultThreadsNo = 3;

        public Service(IPlayerRepository playerRepository, IUserRepository userRepository, IRaceRepository raceRepository, ITeamRepository teamRepository)
        {
            this.playerRepository = playerRepository;
            this.userRepository = userRepository;
            this.raceRepository = raceRepository;
            this.teamRepository = teamRepository;
            loggedClients = new ConcurrentDictionary<int, IObserver>();
        }

        public User FindByUsername(string username)
        {
            var user = userRepository.FindByUsername(username,null);
            if (user == null)
            {
                log.Error("User not found");
                throw new Exception("User not found");
            }
            return user;
        }

        public void Login(User user, IObserver client)
        {
            var userR = userRepository.FindByUsername(user.Username,user.Password);
            if (userR == null)
                throw new MotoException("User not found.");

            if (loggedClients.ContainsKey(userR.Id))
                throw new MotoException("User already logged in.");

            loggedClients[userR.Id] = client;
        }

        public void Logout(User user, IObserver client)
        {
            var usr = userRepository.FindByUsername(user.Username, TextUtils.SimpleDecode(user.Password));
            if (usr != null)
            {
                loggedClients.TryRemove(usr.Id, out _);
                log.Info("User logged out: " + user.Username);
            }
        }

        public void Add(Player player, int raceId, IObserver client)
        {
            try
            {
                var race = raceRepository.FindOne(raceId);
                if (race == null)
                    throw new MotoException("Race not found.");

                player = playerRepository.Save(player);
                if (player == null)
                    throw new MotoException("Failed to save player.");

                if (race.Players == null)
                    race.Players = new List<Player>();

                race.Players.Add(player);
                race.NoPlayers = race.Players.Count;
                raceRepository.Update(race);

                log.Info("Service: Player added");
                Task.Run(() => NotifyPlayerAdded(player));
            }
            catch (Exception e)
            {
                throw new MotoException(e.Message);
            }
        }



        public Team GetTeamByName(string teamName)
        {
            return teamRepository.GetTeamByName(teamName);
        }

        public Race[] GetAllRaces()
        {
            var races = raceRepository.FindAll();
            var raceArray = new Race[races.Count];
            int index = 0;
            foreach (var race in races.Values)
            {
                raceArray[index++] = race;
            }
            return raceArray;
        }

        public Player[] GetRacePlayersOfTeam(int raceId, int teamId)
        {
            var race = raceRepository.FindOne(raceId);
            if (race != null)
            {
                log.Info("Service: Finding players by team");
                var players = new List<Player>();
                foreach (var player in race.Players)
                {
                    if (player.Team == teamId)
                    {
                        players.Add(player);
                    }
                }
                return players.ToArray();
            }
            log.Info("Service: Finding players by team no race found");
            return new Player[0];
        }

        private void NotifyPlayerAdded(Player player)
        {
            var executor = new List<Task>();
            foreach (var client in loggedClients.Values)
            {
                executor.Add(Task.Run(() =>
                {
                    try
                    {
                        client.PlayerAdded(player);
                    }
                    catch (MotoException e)
                    {
                        log.Error("Error notifying player added " + e);
                    }
                }));
            }
            Task.WaitAll(executor.ToArray());
        }
    }
}

File: semestru 4/MotoProject/Server/src/main/java/server/Service.java
package server;


import Services.IObserver;
import Services.IService;
import Services.MotoException;
import model.Player;
import model.Race;
import model.Team;
import model.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repository.*;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import static java.lang.Thread.sleep;

public class Service implements IService {
    private static final Logger log = LogManager.getLogger(Service.class);
    private Map<Integer, IObserver> loggedClients;
    private final PlayerRepository playerRepository;
    private final UserRepository userRepository;
    private final RaceRepository raceRepository;
    private final TeamRepository teamRepository;
    private final int defaultThreadsNo=2;

    public Service(PlayerRepository playerRepository, UserRepository userRepository, RaceRepository raceRepository, TeamRepository teamRepository) {
        this.playerRepository = playerRepository;
        this.userRepository = userRepository;
        this.raceRepository = raceRepository;
        this.teamRepository = teamRepository;
        loggedClients=new ConcurrentHashMap<>();;
    }


    public synchronized void login(User user, IObserver client) throws MotoException {
        Optional<User> usr=userRepository.findByUsername(user.getUsername(),user.getPassword());
        if(usr.isEmpty()){
            throw new MotoException("Username or Password invalid");
        }
        User userR = usr.get();
        if (loggedClients.get(userR.getId()) != null)
            throw new MotoException("User already logged in.");
        loggedClients.put(userR.getId(), client);
    }

    public void logout(User user,IObserver client)  {
        User usr = userRepository.findByUsername(user.getUsername(), TextUtils.simpleDecode(user.getPassword())).get();
        loggedClients.remove(usr.getId());
        log.info("User logged out: " + user.getUsername());
    }

    public synchronized void add(Player player,Integer raceId,IObserver client) throws MotoException {
        try {
            Optional<Race> race = raceRepository.findOne(raceId);
            player = playerRepository.save(player).get();
            race.get().getPlayers().add(player);
            race.get().setNoPlayers(race.get().getPlayers().size());
            raceRepository.update(race.get());
            log.info("Service:Player added");
            for (IObserver c : loggedClients.values()) {
                    try {
                        c.playerAdded(player);
                    } catch (MotoException e) {
                        log.error("Error notifying player added " + e);
                    }
            }
        }catch (Exception e){
            throw new MotoException(e.getMessage());
        }
    }

    public Map<Integer, Player> getPlayersByTeam(Integer teamId) {
        log.info("Service: Finding players by team");
        return playerRepository.getPlayersByTeam(teamId);
    }


    public Optional<Team> getTeamByName(String teamName) {
        return teamRepository.findTeamByName(teamName);
    }

    public synchronized Race[] getAllRaces(){
        return raceRepository.findAll().values().toArray(new Race[0]);
    }

    public synchronized Player[] getRacePlayersOfTeam(Integer raceId, Integer teamId) {
        Optional<Race> race = raceRepository.findOne(raceId);
        if(race.isPresent()) {
            log.info("Service: Finding players by team");
            return race.get().getPlayers().stream()
                    .filter(p -> p.getTeam().equals(teamId))
                    .toList().toArray(new Player[0]);
        }
        log.info("Service: Finding players by team no race found");
        return new Player[0];
    }

private void notifyPlayerAdded(Player player) {
    ExecutorService executor = Executors.newFixedThreadPool(defaultThreadsNo);
    for (IObserver client : loggedClients.values()) {
        executor.execute(() -> {
            try {
                client.playerAdded(player);
            } catch (MotoException e) {
                log.error("Error notifying player added " + e);
            }
        });
    }
    executor.shutdown();
    }
}